src/day10.scala

```diff
@@ -0,0 +1,111 @@
+// using scala 3.1.0
+
+package day10
+
+import scala.util.Using
+import scala.io.Source
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day10"))(_.mkString)
+
+enum CheckResult:
+  case Ok
+  case IllegalClosing(expected: Option[Symbol], found: Symbol)
+  case Incomplete(pending: List[Symbol])
+
+extension (illegalClosing: CheckResult.IllegalClosing)
+  def score: Int = 
+    import Kind.*
+    illegalClosing.found.kind match
+      case Parenthesis => 3
+      case Bracket => 57
+      case Brace => 1197
+      case Diamond => 25137
+
+enum Direction:
+  case Open, Close
+
+enum Kind:
+  case Parenthesis, Bracket, Brace, Diamond
+
+case class Symbol(kind: Kind, direction: Direction):
+  def isOpen: Boolean = direction == Direction.Open
+  def opens(that: Symbol): Boolean = 
+    this.kind == that.kind && 
+      this.direction == Direction.Open &&
+      that.direction == Direction.Close
+    
+def checkChunks(expression: LazyList[Symbol]): CheckResult =
+  @scala.annotation.tailrec
+  def iter(pending: List[Symbol], input: LazyList[Symbol]): CheckResult =
+    if input.isEmpty then
+      if pending.isEmpty then CheckResult.Ok
+      else CheckResult.Incomplete(pending)
+    else
+      val (nextChar #:: remainingChars) = input 
+      if nextChar.isOpen then iter(nextChar :: pending, remainingChars)
+      else pending match
+        case Nil => CheckResult.IllegalClosing(None, nextChar)
+        case lastOpened :: previouslyOpened =>
+          if lastOpened.opens(nextChar) then iter(previouslyOpened, remainingChars)
```

Comment:
Here we already know that `lastOpened`’s direction is `Open`, and that `nextChar`’s direction is `Close`. So, we only need to check whether `lastOpened.kind == nextChar.kind` (instead of calling `lastOpened.opens(nextChar)`).